pkg/reconciliation/reconcile_racks.go

```diff
@@ -1806,97 +1809,99 @@ func (rc *ReconciliationContext) startCassandra(endpointData httphelper.CassMeta
 	return rc.labelServerPodStarting(pod)
 }
 
-// returns the name of one rack without any ready node
-func (rc *ReconciliationContext) startOneNodePerRack(endpointData httphelper.CassMetadataEndpoints, readySeeds int) (string, error) {
-
+// startOneNodePerRack starts the first pod in each rack in a deterministic order, dictated by the
+// order in which the racks were declared in the datacenter spec. It returns true if a Cassandra
+// node is not ready yet, and an error if that node also failed to start.
+func (rc *ReconciliationContext) startOneNodePerRack(endpointData httphelper.CassMetadataEndpoints, readySeeds int) (notReady bool, err error) {
 	rc.ReqLogger.Info("reconcile_racks::startOneNodePerRack")
 
-	rackReadyCount := map[string]int{}
-	for _, rackInfo := range rc.desiredRackInformation {
-		rackReadyCount[rackInfo.RackName] = 0
-	}
+	labelSeedBeforeStart := readySeeds == 0 && !rc.hasAdditionalSeeds()
 
-	for _, pod := range rc.dcPods {
-		rackName := pod.Labels[api.RackLabel]
-		if isServerReady(pod) {
-			rackReadyCount[rackName]++
-		}
-	}
+	for _, statefulSet := range rc.statefulSets {
 
-	// if the DC has no ready seeds, label a pod as a seed before we start Cassandra on it
-	// and also consider additional seeds
+		pod := rc.getDCPodByName(statefulSet.Name + "-0")
+		notReady, err = rc.startNode(pod, labelSeedBeforeStart, endpointData)
+		if notReady {
+			break
+		}
 
-	externalSeedPoints := 0
-	if existingEndpoints, err := rc.GetAdditionalSeedEndpoint(); err == nil {
-		externalSeedPoints = len(existingEndpoints.Subsets[0].Addresses)
 	}
+	return
+}
+
+// startAllNodes starts all nodes in the datacenter in a deterministic order, dictated by the order
+// in which the racks were declared in the datacenter spec, and in such a way that at any given
+// time, the number of started nodes in a rack cannot be greater or lesser than the number of
+// started nodes in all other racks by more than 1. It returns true if a Cassandra node is not ready
+// yet, and an error if that node also failed to start.
+func (rc *ReconciliationContext) startAllNodes(endpointData httphelper.CassMetadataEndpoints) (notReady bool, err error) {
+	rc.ReqLogger.Info("reconcile_racks::startAllNodes")
 
-	labelSeedBeforeStart := readySeeds == 0 && len(rc.Datacenter.Spec.AdditionalSeeds) == 0 && externalSeedPoints == 0
+	for podRankWithinRack := 0; ; podRankWithinRack++ {
```

Comment:
Out of interest, would you explain the benefit vs, lets say following algorithm (the problem here looks very much like a dynamic programming)

```go
smallestRack := 0
for i := 0; i < len(rc.statefulSets); i++ {
    if rc.statefulSets[i].Status.ReadyReplicas < rc.statefulSets[smallestRack].Status.ReadyReplicas {
        smallestRack = i
    }
}

fmt.Printf("Next rack to start is: %d\n", smallestRack)
```

Also, is .Replicas really the number you want to compare to vs. ReadyReplicas?

Reply:
First, my understanding is that we want here to use the *desired* number of pods for each sts, thus we should use `.Replicas`.

Second, switching to `.ReadyReplicas` could indeed simplify the code, but can we guarantee that if a pod is counted as a ready replica, it is not only ready, but also started and has joined the ring? Because if that's not the case, we should NOT keep bootstrapping the next pods.

Reply:
How would Replicas help with that behavior? Isn't .Spec.Replicas updated when we add the pods? They don't have to be ready at all. Our startup process should take care of preventing next startup (there's one pod that's Started-but-not-Ready) before starting next ones. Though for that reason, perhaps it doesn't matter which one we use, since when we get to this point, we should have no starting pods anymore.

Reply:
So, what should I do here? Should I change as you suggested above?

Reply:
No, not necessarily. I only want to ensure that whoever reads it next can easily see what it's doing (and so avoid anything that looks pretty and complex). Would you think from that perspective the code is self-explanatory enough (the comment on the method only describes what we want to achieve, not how we achieve it) ?

Reply:
To me, it is self-explanatory as it is now, however that's a subjective notion. You are never going to get everybody on the same page on such matters.